// @mitm/assetbundlecompiler/dist/src/build_targets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StandaloneOSX = 'StandaloneOSXUniversal';
exports.StandaloneOSXUniversal = 'StandaloneOSXUniversal'; // @UnityVersion remove when 5.x support is dropped
exports.StandaloneOSXIntel = 'StandaloneOSXIntel'; // @UnityVersion remove when 5.x support is dropped
exports.StandaloneWindows = 'StandaloneWindows';
exports.iOS = 'iOS';
exports.Android = 'Android';
exports.StandaloneLinux = 'StandaloneLinux';
exports.StandaloneWindows64 = 'StandaloneWindows64';
exports.WebGL = 'WebGL';
exports.WSAPlayer = 'WSAPlayer';
exports.StandaloneLinux64 = 'StandaloneLinux64';
exports.StandaloneLinuxUniversal = 'StandaloneLinuxUniversal';
exports.StandaloneOSXIntel64 = 'StandaloneOSXIntel64'; // @UnityVersion remove when 5.x support is dropped
exports.Tizen = 'Tizen';
exports.PSP2 = 'PSP2';
exports.PS4 = 'PS4';
exports.XboxOne = 'XboxOne';
exports.SamsungTV = 'SamsungTV'; // @UnityVersion remove when 5.x support is dropped
exports.N3DS = 'N3DS';
exports.WiiU = 'WiiU';
exports.tvOS = 'tvOS';
exports.Switch = 'Switch';
